### PR TITLE
[query] add TableIR support to requiredness analysis

### PR DESCRIPTION
**Pull request by catoverdrive**
Adds support for TableIR nodes, as well as some test coverage for all supported TableIR nodes.

I think I'm going to choose not to support RelationalLetTable at this point; I think we lift relational lets before we lower MatrixTables, so we would never need to support it currently in our normal lowering pipeline.

### CODE REVIEW

**Comment by tpoterba** [comment]
> I think I'm going to choose not to support RelationalLetTable at this point; I think we lift relational lets before we lower MatrixTables, so we would never need to support it currently in our normal lowering pipeline.


This isn't true -- we generate relational lets in the MatrixLowering pipeline!

File: hail/src/main/scala/is/hail/expr/ir/Requiredness.scala [hunk 1]
```diff
@@ -176,15 +190,157 @@ class Requiredness(val usesAndDefs: UsesAndDefs, ctx: ExecuteContext) {
...
+      case TableFilter(child, _) => requiredness.unionFrom(lookup(child))
+      case TableHead(child, _) => requiredness.unionFrom(lookup(child))
+      case TableTail(child, _) => requiredness.unionFrom(lookup(child))
+      case TableRepartition(child, n, strategy) => requiredness.unionFrom(lookup(child))
+      case TableDistinct(child) => requiredness.unionFrom(lookup(child))
+      case TableOrderBy(child, sortFields) => requiredness.unionFrom(lookup(child))
+      case TableRename(child, rMap, gMap) => requiredness.unionFrom(lookup(child))
+      case TableFilterIntervals(child, intervals, keep) => requiredness.unionFrom(lookup(child))
+      case RelationalLetTable(name, value, body) =>
+        fatal(s"RelationalLetTable must be lifted prior to this pass \n${ Pretty(node) }")
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Requiredness.scala:241]: Let's do that in a followup PR

File: hail/src/main/scala/is/hail/expr/types/TypeWithRequiredness.scala [hunk 2]
```diff
@@ -60,6 +60,20 @@ sealed abstract class BaseTypeWithRequiredness {
...
 
+  protected[this] def _unionWithIntersection(ts: Seq[BaseTypeWithRequiredness]): Unit = {
+    var i = 0
+    while(i < children.length) {
+      children(i).unionWithIntersection(ts.map(_.children(i)))
+      i += 1
+    }
+  }
+  def unionWithIntersection(ts: Seq[BaseTypeWithRequiredness]): Unit = {
+    println(s"$this: ${ ts.mkString(",") }")
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/types/TypeWithRequiredness.scala:71]: prints

File: hail/src/main/scala/is/hail/expr/ir/Requiredness.scala [hunk 3]
```diff
@@ -176,15 +190,157 @@ class Requiredness(val usesAndDefs: UsesAndDefs, ctx: ExecuteContext) {
...
+    node match {
+      //statically known
+      case TableLiteral(typ, rvd, enc, encodedGlobals) =>
+        requiredness.rowType.fromPType(rvd.rowPType)
+        requiredness.globalType.fromPType(enc.encodedType.decodedPType(typ.globalType))
+      case TableRead(typ, dropRows, tr) =>
+        val (rowPType, globalPType) = tr.rowAndGlobalPTypes(ctx, typ)
+        requiredness.rowType.fromPType(rowPType)
+        requiredness.globalType.fromPType(globalPType)
+      case TableRange(_, _) =>
```
[the remark this answers is not in the document]
Comment (catoverdrive) [reply hail/src/main/scala/is/hail/expr/ir/Requiredness.scala:228]: nope---TableRange types are all required.

File: hail/src/main/scala/is/hail/expr/ir/Requiredness.scala [hunk 4]
```diff
@@ -260,15 +272,160 @@ class Requiredness(val usesAndDefs: UsesAndDefs, ctx: ExecuteContext) {
...
+        addTableBinding(child)
+      case TableMapGlobals(child, newGlobals) =>
+        addTableBinding(child)
+      case TableKeyByAndAggregate(child, expr, newKey, nPartitions, bufferSize) =>
+        addTableBinding(child)
+      case TableAggregateByKey(child, expr) =>
+        addTableBinding(child)
+      case _ =>
+        usesAndDefs.uses(node).foreach { u =>
+          println(u.t)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/Requiredness.scala:290]: stray println

**Pull request merged**